### ISSUE TITLE
false positive "name not defined" with list comprehension in function arguments default value

### ISSUE THREAD

**Issue by perrinjerome** [python/mypy#9346]
* Are you reporting a bug, or opening a feature request?

This is a bug I believe 

* Please insert below the code you are checking with mypy

```python
def a():
  pass

def b(
    bb=[list_comprehension_argument for list_comprehension_argument in "abc" 
          if list_comprehension_argument]
  ):
  pass
```

* What is the actual behavior/output?

```
mypy repro.py
repro.py:5: error: Name 'list_comprehension_argument' is not defined
repro.py:6: error: Name 'list_comprehension_argument' is not defined
Found 2 errors in 1 file (checked 1 source file)
```

* What is the behavior/output you expect?

I believe there should be no error reported.

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

this is on mypy from git master, `mypy 0.790+dev.2b704df25ca299e6936927be8d41d2ab945cee2a`

* What are the mypy flags you are using? (For example --strict-optional)

This is using the default flags.

----

This only seem to happens if there's another function like this one, *before* the function with list comprehension in arguments:
```python
def a():
  pass
```
When removing this other function or putting it *after* the one with list comprehension in argument, this false positive is not reported, I mean no error are reported for:
```python
def b(
    bb=[list_comprehension_argument for list_comprehension_argument in "abc" 
          if list_comprehension_argument]
  ):
  pass

def a():
  pass
```

or 

```python
def b(
    bb=[list_comprehension_argument for list_comprehension_argument in "abc" 
          if list_comprehension_argument]
  ):
  pass
```

**Comment by gvanrossum** [comment]
Thanks for reporting this! I've confirmed everything you said -- it doesn't occur if `def a()...` is removed or moved after `def b()`. (IOW, it doesn't happen if `b()` is the first function in the file.)

I haven't looked through the mypy source code, but it would appear that mypy isn't properly managing some scoping-related internal datastructure.

For posterity, here's a shorter repro:
```
def a():  # The error disappears if this function is removed
    pass

def b(bb=[x for x in "abc"]):  # E: Name 'x' is not defined
    pass
```

@JukkaL Any idea?

**Comment by perrinjerome** [comment]
Also, this was working before, it seems to be a regression from c610a312853921d09083f7525c183810e4232056

**Comment by gvanrossum** [comment]
Thanks very much! @JelleZijlstra can you have a look?

**Comment by JelleZijlstra** [comment]
I'm guessing it's a result of https://github.com/python/mypy/commit/c610a312853921d09083f7525c183810e4232056#diff-ebbe3135b7d21c5fa8783a4d0d0ccd48R525. Previously we just didn't type check argument defaults at all, so we didn't get errors either. I'm not sure how to explain the effects of function ordering reported by the OP though.

**Comment by gvanrossum** [comment]
Possibly some manipulation of the comprehension stack is needed around that visit() call?

**Comment by matheusfillipe** [comment]
Any updates on this? I also noticed it doesn't happen if there are 

```python
from typing import TypedDict


class SomeClass:
    pass

class AnotherClass(TypedDict):
    thing: str

def somefunc(
    someparam: tuple = tuple(number for number in range(10)),
    ):
    pass
    
```

That works even if it is a subclass which the parent has methods defined, but when you defined your own method it fails:
```python
class SomeClass:
    def anymethod(self):
        pass

def somefunc(
    someparam: tuple = tuple(number for number in range(10)),
    ):
    pass
```

Fails with:
```
demo.py:6: error: Name "number" is not defined  [name-defined]
Found 1 error in 1 file (checked 1 source file)
```

Also doesn't matter what type of comprehension it is, list, dict, generator, the false positive will happen.
